### PR TITLE
Revjtanton/issue 63

### LINKED ISSUE #31: Conditional Dockerhub deploy
### User Story
As a developer 
I'd like to push the backend docker image to dockerhub
So that I can pull it into the frontend docker-compose setup 

### Acceptance Criteria
- [ ] Deploy to Dockerhub on merge to master, not PR 
- [ ] Look into more secure credential store process

### Notes
https://docs.docker.com/engine/reference/commandline/login/#credentials-store

### LINKED ISSUE #63: Installation Error
api_1  | [nodemon] 2.0.2
api_1  | [nodemon] to restart at any time, enter `rs`
api_1  | [nodemon] watching dir(s): *.*
api_1  | [nodemon] watching extensions: js,mjs,json
api_1  | [nodemon] starting `babel-node src/index.js`
api_1  | Executing (default): SELECT 1+1 AS result
api_1  | Executing (default): CREATE TABLE IF NOT EXISTS "public"."Contacts" ("id" UUID UNIQUE , "UserId" VARCHAR(255) REFERENCES "public"."User" ("id") ON DELETE NO ACTION ON UPDATE CASCADE, "name" VARCHAR(255), "phone" JSON, "email" JSON, "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL, "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL, PRIMARY KEY ("id"));
api_1  | Connection to database successful
*db_1   | 2020-03-21 08:40:07.024 UTC [32] ERROR:  relation "public.User" does not exist*
db_1   | 2020-03-21 08:40:07.024 UTC [32] STATEMENT:  CREATE TABLE IF NOT EXISTS "public"."Contacts" ("id" UUID UNIQUE , "UserId" VARCHAR(255) REFERENCES "public"."User" ("id") ON DELETE NO ACTION ON UPDATE CASCADE, "name" VARCHAR(255), "phone" JSON, "email" JSON, "createdAt" TIMESTAMP WITH TIME ZONE NOT NULL, "updatedAt" TIMESTAMP WITH TIME ZONE NOT NULL, PRIMARY KEY ("id"));
api_1  | Unhandled rejection SequelizeDatabaseError: relation "public.User" does not exist
api_1  |     at Query.formatError (/usr/src/node_modules/sequelize/lib/dialects/postgres/query.js:366:16)
api_1  |     at /usr/src/node_modules/sequelize/lib/dialects/postgres/query.js:72:18
api_1  |     at tryCatcher (/usr/src/node_modules/bluebird/js/release/util.js:16:23)
api_1  |     at Promise._settlePromiseFromHandler (/usr/src/node_modules/bluebird/js/release/promise.js:547:31)
api_1  |     at Promise._settlePromise (/usr/src/node_modules/bluebird/js/release/promise.js:604:18)
api_1  |     at Promise._settlePromise0 (/usr/src/node_modules/bluebird/js/release/promise.js:649:10)
api_1  |     at Promise._settlePromises (/usr/src/node_modules/bluebird/js/release/promise.js:725:18)
api_1  |     at _drainQueueStep (/usr/src/node_modules/bluebird/js/release/async.js:93:12)
api_1  |     at _drainQueue (/usr/src/node_modules/bluebird/js/release/async.js:86:9)
api_1  |     at Async._drainQueues (/usr/src/node_modules/bluebird/js/release/async.js:102:5)
api_1  |     at Immediate.Async.drainQueues [as _onImmediate] (/usr/src/node_modules/bluebird/js/release/async.js:15:14)
api_1  |     at processImmediate (internal/timers.js:439:21)
api_1  |     at process.topLevelDomainCallback (domain.js:131:23)
api_1  | [nodemon] clean exit - waiting for changes before restart

While trying to run with `docker-compose up`. I get the following error. Anything I'm missing on doing. Not a nodejs fellow

### PR DESCRIPTION
**Pull request by revjtanton**
This resolves the Docker implementation problems we have been having and sets up Docker deployments.
resolves #63 
resolves #31 

## Todos

- [x] Dockerhub setup
- [x] Documentation
- [x] Fixing docker-compose variables issue

## Deploy Notes

It is very important that you've already built the Dockerfile before running docker-compose because we are prepping for Dockerhub integration so it will not build automatically yet. 

It is also very important that you properly setup your `.env` file. This is thoroughly outlined in the README.

## Impacted Areas in Application

List general components of the application that this PR will affect:

- Docker

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,38 +35,31 @@ To work on this project you should have:
...
-DATABASE_USER=<your database user>
-DATABASE_PASSWORD=<your database password>
 DATABASE_SCHEMA=<your database schema>
 JWT_KEY=<your secret JWT seed phrase or key>
 DATABASE_URL_DEV=<your connection string based on above variables>
 ' >> ./.env
 ```
 
-And example of the `DATABASE_URL_DEV` would be `DATABASE_URL_DEV=postgres://postgres:.@localhost:5432/postgres`
+And example of the `DATABASE_URL` would be `DATABASE_URL=postgres://user:pass@example.com:5432/dbname`
```
Comment (jbixon13) [inline README.md]: DATABASE_URL_DEV in the prior .env section and DATABASE_URL here. Additionally, there are no variables associated with user, pass, or dbname in the prior section (as they were removed), are they meant to be defined for the first time here? This seems unclear if someone hasn't seen these variables declared previously.

File: README.md [hunk 2]
```diff
@@ -63,53 +70,56 @@ _We do not recommend using the default options for PostgreSQL. The above values
 - `DATABASE_HOST`: The IP address Docker is running on. You can find this by running `docker-machine ip` but it's usually `192.168.99.100` instead of `localhost`
 - `DATABASE_URL_DEV`: This will need to be adjusted as well, for example `DATABASE_URL_DEV=postgres://postgres:.@localhost:5432/postgres` would become `DATABASE_URL_DEV=postgres://postgres:.@192.168.99.100:5432/postgres`
```
Comment (jbixon13) [inline README.md]: Should these be DATABASE_URL_DEV or DATABASE_URL? Will depend on answers to prior comments.

File: README.md [hunk 3]
```diff
@@ -63,53 +70,56 @@ _We do not recommend using the default options for PostgreSQL. The above values
...
 If you're running a database in another way then we trust you can sort it out on your own because you're awesome :sunglasses:
 
-### Sequelize _<optional>_
-You can run the application without doing anything and it will create the tables needed to operate automatically. It will not, however, create users. If you would like to seed your database with users you will need to follow a few steps.
-1. You may create your database tables without running the application by running `npm run db-create`.
-2. You can now seed your database by running `npm run db-seed`. 
+### 2.3.1. Sequelize
+To properly start the application the database needs to be built by Sequlize ahead of time. To do that run the following commands
+1. You must create your database tables without running the application by running `npm run db-create` first.
+2. _optional_ You can now seed your database if you wish by running `npm run db-seed`.
```
Comment (jbixon13) [inline README.md]: I'm getting an error on npm run db-create: 'Dialect needs to be explicitly supplied as of v4.0.0'

Is this due to Postgres not being defined where it was previously (.env?, Dockerfile?)

File: README.md (hunk 2 above)
Comment (revjtanton) [reply README.md]: It should be DATABASE_URL. I will fix this in another PR.

File: README.md (hunk 3 above)
Comment (revjtanton) [reply README.md]: You have to have your `DATABASE_URL` defined in the `.env` file. The URL contains the un/pw, host, db, etc.

**Pull request merged**